design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,296 @@
+that uses this API.
+
+This KLIP proposes that we:
+
+* Create a ksqlDB client (initially in Java and JavaScript) as the primary interaction point to the event streaming platform
+for an application
+* Create a new HTTP2 based server API using Vert.x that supports the client and enables the functionality of the client to be delivered in a straightforward
+and efficient way
+* Migrate the functionality of the current HTTP/REST API over to the new server implementation and retire the parts that are no longer needed.
+
+## What is in scope
+
+### The client
+
+* We will create a new client, initially in Java (and potentially in JavaScript). Clients for other languages such as Python and Go will follow later.
+* The client will initially support execution and streaming of queries (both push and pull), inserting of rows into streams, DDL operations and admin operations
+such as list and describe.
+* The client will support a reactive / async interaction style using Java CompletableFuture and Reactive Streams / JDK9 Flow API (or a copy thereof if we cannot upgrade from Java 8)
+* The client will also support a direct / synchronous interaction style.
+* The server API will be very simple, based on HTTP2 with a simple text/json based encoding therefore it will be very simple to use directly from vanilla
+HTTP2 clients if a ksqldb client is not available
+
+### The Server API
+
+We will create a new, simple HTTP2 API for streaming query results from server to client and
+for streaming inserts from client to server. We will use a simple text/JSON encoding for data.
+Please note, this is not a REST API, it's a streaming HTTP API. The API does not follow REST principles. REST is inherently designed for request/response (RPC)
+style interactions, not streaming.
+
+The API will have the following characteristics:
+
+* Multiplexed (because of HTTP2 multiplexing)
+* Back-pressure (because of HTTP2 flow control)
+* Text based so easy to use and view results using command line tools such as curl
+* Can be used from any vanilla HTTP2 client for any language
+* Simple newline based delimitation so easy to parse results on client side
+* JSON encoding of data so very easy to parse as most languages have good JSON support
+* CORS support
+* HTTP basic auth support
+* TLS
+
+#### Query streaming
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/query-stream" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added here for the sake of clarity but the real JSON must not contain newlines)
+
+````
+{
+"query": "select * from foo", <----- the SQL of the query to execute
+"push": true,                 <----- if true then push query, else pull query
+"limit": 10                   <----  If specified return at most this many rows,
```

Comment:
This was meant as an illustration, not an exact definition of the arguments to the real endpoint. (I think there is a comment in there already to that effect ;) )

Reply:
That makes sense @purplefox.  Otherwise I would have asked why there needs to be a `"push"` field, as the behavior of push vs. pull should be determined by the query itself (cf. `"query`"), not by an out-of-band parameter. :-)